### PR TITLE
Cls tests

### LINKED ISSUE #97: Missing tests: cl.py
- [x] workspace for spin-0 approximation in covariance (maybe this is part of the covariance unit tests)
- [x] As above, include the possibility of reading it from file.
- [x] Test an error is thrown when reading a cl file that is incompatible (line 178)
- [x] Test `get_bandpower_windows`
- [x] Test CCL calculation for cosmic shear and cmb convergence
- [x] Test `read_symmetric` in line 278

### LINKED ISSUE #98: Missing tests: cov.py
- [x] Test for spin-0 covariance
- [x] Test non-overlaping datasets gives you a zero covariance
- [x] Test multiplicative bias marginalization

### PR DESCRIPTION
**Pull request by damonge**
Closes #97 
Closes #98

### CODE REVIEW

**Comment by damonge** [comment]
We just hit 90%!
@carlosggarcia this is ready for review

File: xcell/tests/test_cls.py [hunk 1]
```diff
@@ -33,9 +37,11 @@ def get_config(fsky=0.2, fsky2=0.3):
         'baryons_power_spectrum': 'nobaryons',
     }
     dummy0 = {'mask_name': 'mask_dummy0', 'mapper_class': 'MapperDummy',
-              'cosmo': cosmo, 'nside': nside, 'fsky': fsky, 'seed': 0}
+              'cosmo': cosmo, 'nside': nside, 'fsky': fsky, 'seed': 0,
+              'dtype': dtype}
     dummy1 = {'mask_name': 'mask_dummy1', 'mapper_class': 'MapperDummy',
-              'cosmo': cosmo, 'nside': nside, 'fsky': fsky2, 'seed': 100}
+              'cosmo': cosmo, 'nside': nside, 'fsky': fsky2, 'seed': 100,
+              'dtype': dtype2}
```
Comment (carlosggarcia) [inline xcell/tests/test_cls.py:44]: After getting confused lower in this file, I'd rename `dtype` and `dtype2` to `dtype0` and `dtype1` as the dummy mappers.

File: xcell/tests/test_cls.py [hunk 2]
```diff
@@ -203,3 +248,150 @@ def test_symmetric():
...
+                f'Dummy__{perm[0]}',
+                f'Dummy__{perm[1]}',
+                f'Dummy__{perm[2]}',
+                f'Dummy__{perm[3]}')
+    assert covc0.spin0
+    cov0 = covc0.get_covariance()
+    assert cov0.shape == (ncls1*nbpw, ncls2*nbpw)
+    shutil.rmtree(tmpdir1)
+
+    if ncls1 == ncls2:
```
Comment (carlosggarcia) [inline xcell/tests/test_cls.py:346]: A bit confused here. Is this only `True` when checking the shear-shear x shear-shear case?

File: xcell/tests/test_cls.py [hunk 3]
```diff
@@ -203,3 +248,150 @@ def test_symmetric():
...
+        # Check that they are the same on all bandpowers
+        # except the first one
+        r = np.diag(cov2)/np.diag(cov0)-1
+        # This loops through EE, EB, BE, BB
+        for i in range(ncls1):
+            assert np.all(r[i::ncls1][1:] < 1E-5)
```
Comment (carlosggarcia) [inline xcell/tests/test_cls.py:352]: Why is this `True`?

**Review by carlosggarcia — CHANGES_REQUESTED** [review]
Minor comments and some questions

File: xcell/tests/test_cls.py (hunk 3 above)
Comment (damonge) [reply xcell/tests/test_cls.py:352]: in the full sky, the spin-0 and general approximations should be almost the same (except for the first bandpower).

File: xcell/tests/test_cls.py (hunk 2 above)
Comment (damonge) [reply xcell/tests/test_cls.py:346]: for 02x02 and 22x22
in all other cases most of the covariance elements are zero, so it's hard to compare those generically and I didn't think it was worth it

File: xcell/tests/test_cls.py [hunk 4]
```diff
@@ -127,7 +164,7 @@ def test_get_covariance():
     cov = cov_class.get_covariance()
 
     cov_m = np.zeros_like(cov)
-    diag = (2 * cl_m1 ** 2) / (2 * ell + 1)
+    diag = (2 * cl_m1 ** 2) / (2 * ell + 1) / 4
```
[the remark this answers is not in the document]
Comment (damonge) [reply xcell/tests/test_cls.py:167]: our bandpowers have width 4 (d_ell), so the covariance for each bandpower scales inversely with that.

File: xcell/tests/test_cls.py (hunk 3 above)
Comment (carlosggarcia) [reply xcell/tests/test_cls.py:352]: Oooh! I hadn't noticed that `fsky=1`, sorry!

File: xcell/tests/test_cls.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carlosggarcia) [reply xcell/tests/test_cls.py:167]: Make sense. Thanks!

File: xcell/tests/test_cls.py (hunk 2 above)
Comment (carlosggarcia) [reply xcell/tests/test_cls.py:346]: Makes sense

File: xcell/tests/test_cls.py [hunk 5]
```diff
@@ -67,6 +73,19 @@ def get_cov_class(fsky=0.2):
...
     get_cl_class()
     get_cov_class()
+    shutil.rmtree(tmpdir1)
+
+
+def test_get_nmtbin():
+    data = get_config()
+    cl1 = Cl(data, 'Dummy__0', 'Dummy__0')
+    shutil.rmtree(tmpdir1)
+    data['cls']['Dummy-Dummy']['bpw_edges'] = data.pop('bpw_edges')
```
[the remark this answers is not in the document]
Comment (damonge) [reply xcell/tests/test_cls.py:88]: Done

File: xcell/tests/test_cls.py (hunk 1 above)
Comment (damonge) [reply xcell/tests/test_cls.py]: Done

**Comment by damonge** [comment]
All done, back to you @carlosggarcia !

**Review by carlosggarcia — APPROVED** [review]

**Pull request merged**